### PR TITLE
Feature/78 add fairness metrics

### LINKED ISSUE #78: Add fairness metrics
### Problem

Is your feature request related to a problem?  No
Add some metrics concerning the fairness of the datasets and of the models

## Concerned template

  - [x] NLP template
  - [x] NUM template
  - [ ] VISION template
  - [ ] How templates are generated - Jinja

### Solution

MVP : Implements the libraries fairlearn (https://github.com/fairlearn/fairlearn) and fairlens (https://github.com/synthesized-io/fairlens) in gabarit.

Possible additions : Add custom metrics to fairlearn and fairlens

### PR DESCRIPTION
**Pull request by gsolard**
Add some metrics concerning the fairness of the datasets and of the models

## ✒️ Context

Add some metrics concerning the fairness of the datasets and of the models

- What kind of change does this PR introduce ?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧱 Description of Changes

- Added a script to generate files containing the fairness metrics

  - [ ] This is a change for the NLP template
  - [x] This is a change for the NUM template
  - [ ] This is a change for the VISION template
  - [ ] This changes how templates are generated (i.e. a Jinja change)

## 🩺 Testing

- Test the fairness script

  - [ ] This change does not need new tests
  - [x] Added/Updated unit tests
  - [x] Added/Updated functionals tests (i.e. e2e)

## 🔗 References

- **Issue**: Closes #78

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the GNU AFFERO GENERAL PUBLIC LICENSE.

### CODE REVIEW

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+def get_metrics_functions(data_col: pd.Series)-> dict:
+    '''Gets the metrics to calculate using the target and the prediction
+    
+    Args:
+        data_col (pd.Series) : The data on which we want to infer the metrics
+    Returns:
+        dict : the dictionary containing the metrics to calculate between the target and the predictions
+    '''
+    distr_type = fl_utils.infer_distr_type(data_col)
+    metrics_functions = {"count":fairlearn.metrics.count}
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:170]: pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+
+
+def plot_count(df_metrics: pd.DataFrame) -> matplotlib.axes._axes.Axes:
+    '''Plots the count pie from a df_metrics obtained from a MetricFrame
+
+    Args:
+        df_metrics (pd.DataFrame) : The dataframe containing a column count and the indices corresponding to the groups
+    Returns:
+        matplotlib.axes._axes.Axes : The matplotlib axes for the figure
+
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:223]: Space

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+    Returns:
+        matplotlib.axes._axes.Axes : The matplotlib axes for the figure
+    
+    '''
+    fig = df_metrics[['count']].plot(kind="pie",
+                         subplots=True,
+                         layout=[1,1],
+                         legend=False,
+                         figsize=[12,8],
+                         title='Size of each sensitive subgroup')
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:230]: Can this be less lines ? + PEP8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+
+def plot_one_metric(df_metrics: pd.DataFrame, metric_name: str) -> matplotlib.axes._axes.Axes:
+    '''Plots the bar graph for the chosen metric.
+
+    Args:
+        df_metrics (pd.DataFrame) : The dataframe containing a column with the metric for each group
+        metric_name (str) : The name of the chosen metric
+    Returns:
+        matplotlib.axes._axes.Axes : The matplotlib axes for the figure
+
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:243]: Space

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    
+    '''
+    ordered_df = df_metrics[[metric_name]].sort_values(metric_name)
+    index_overall = list(ordered_df.index).index(('overall',)*len(df_metrics.index[0]))
+    fig = ordered_df.plot(kind="bar",
+                         subplots=True,
+                         layout=[1,1],
+                         legend=False,
+                         figsize=[12,8],
+                         title=metric_name+' for each sensitive subgroup')
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:252]: Less lines ? + PEP8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+    ax = fig[0][0]
+    # Color in orange the bar corresponding to the whole dataset
+    for i, patch in enumerate(ax.patches):
+        if i==index_overall:
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:256]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+    '''Saves the graphs given a metric_frame and returns the dataframe used to obtain them.
+
+    Args:
+        metric_frame (fairlearn.metrics.MetricFrame) : The MetricFrame from which we want to plot the graphs
+        output_path (str) : The path where to save the plots
+    Kwargs:
+        sep (str): Separator to use with the .csv files
+        encoding (str): Encoding to use with the .csv files
+        mlflow_logger (MLflowLogger) : The logger to log the metrics in MLflow
+
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:272]: Space

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+        mlflow_logger (MLflowLogger) : The logger to log the metrics in MLflow
+    
+    '''
+    df_metrics = metric_frame.by_group
+    # Pie plot for count
+    ax_count = plot_count(df_metrics)
+    ax_count.figure.savefig(os.path.join(output_path, 'fairness_count_groups.png'), bbox_inches="tight")
+    # Add the 'overall' row to df_metrics
+    nb_level_index = len(df_metrics.index[0])
+    df_overall =  pd.DataFrame(metric_frame._overall).T
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:280]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+    '''Gives the 'next' date according to a step. For example, if we are the 24th of March 2021, the 'next' date with a step
+    'year' is the 1st of January 2022, with a step 'month', it is the 1st of April 2021, with a step 'day', it is the 25th of 
+    March 2021...
+    
+    Args:
+        date (pd.Timestamp): The date to consider
+        step (str) : The step to consider (in ['year', 'month', 'day', 'hour', 'minute', 'second'])
+    Returns:
+        (datetime.datetime) : The next date to consider
+
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:331]: Space

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+        new_date = date.replace(day=1)+datetime.timedelta(days=32)
+        new_date = datetime.datetime(new_date.year, new_date.month, 1)
+    if step == 'day':
+        new_date = date.replace(hour=0)+datetime.timedelta(seconds=86401)
+        new_date = datetime.datetime(new_date.year, new_date.month, new_date.day)
+    if step == 'hour':
+        new_date = date.replace(minute=0)+datetime.timedelta(seconds=3601)
+        new_date = datetime.datetime(new_date.year, new_date.month, new_date.day, new_date.hour)
+    if step == 'minute':
+        new_date = date.replace(second=0)+datetime.timedelta(seconds=61)
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:346]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+        new_date = datetime.datetime(new_date.year, new_date.month, new_date.day, new_date.hour, new_date.minute)
+    if step == 'second':
+        new_date = date+datetime.timedelta(seconds=1)
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:349]: pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+    Returns:
+        List[str] : A list of labels to name the bins
+    '''
+    # The list of the dates for the left limits of the bins
+    list_begin_date = bins[:-1]
+    # The list of the dates for the right limits of the bins
+    if step != 'second':
+        list_end_date = [get_previous_date(date-datetime.timedelta(seconds=1), step) for date in bins[1:]]
+    else:
+        list_end_date = [date-datetime.timedelta(seconds=1) for date in bins[1:-1]]+[bins[-1]]
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:394]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 13]
```diff
@@ -0,0 +1,535 @@
...
+    date_format = ''
+    list_steps = ['year', 'month', 'day', 'hour', 'minute', 'second']
+    list_format = ['/%Y', '/%m', '/%d', '|%H', ':%M', ':%S']
+    # Max step to consider (ie the 'precision' we want)
+    max_index = list_steps.index(step)+1
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:399]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 14]
```diff
@@ -0,0 +1,535 @@
...
+    # Checks if we already added something to the date_format (in order not to 'hop' over a step)
+    test_begin = False
+    # We loop on the possible steps and add it to the date_format if necessary
+    for attribute, split_format in zip(list_steps[:max_index], list_format[:max_index]):
+        if not len({getattr(date, attribute) for date in list_begin_date+list_end_date})==1 or test_begin:
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:404]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 15]
```diff
@@ -0,0 +1,535 @@
...
+                date_format += split_format
+    # Now that the date_format as been calculated, actually transforms the limits in str
+    labels = []
+    for date_begin, date_end in zip(list_begin_date, list_end_date):
+        label_begin = date_begin.strftime(date_format)
+        label_end = date_end.strftime(date_format)
+        if label_begin == label_end:
+            labels.append(label_begin)
+        else:
+            labels.append(label_begin+'-'+label_end)
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:418]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 16]
```diff
@@ -0,0 +1,535 @@
...
+    labels = [prefix_label+label for label in labels]
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:419]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 17]
```diff
@@ -0,0 +1,535 @@
...
+        date_bins (List[pd.Timestamp]) : The limits to the bins (as given by a pd.qcut for example)
+        prefix_label (str) : A prefix to add to each category label
+        ratio_min (float) : The minimum ratio wanted between the less populated category over the most populated
+    Returns:
+        pd.Series : The binned column
+    '''
+    # For each 'step' ie 'precision'
+    for step in ['year', 'month', 'day', 'hour', 'minute']:
+        # Gets new 'natural' bins and sees if they verify the ratio requirement
+        new_bins = [get_previous_date(date_bins[0], step)]+[get_next_date(date, step) for date in date_bins[1:]]
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:439]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 18]
```diff
@@ -0,0 +1,535 @@
...
+        if len(set(new_bins)) == len(date_bins):
+            labels = get_labels_from_bins(new_bins, step, prefix_label)
+            new_col, test_ratio = check_sufficient_balance(date_column, new_bins, labels=labels, ratio_min=ratio_min)
+            if test_ratio:
+                return new_col
+        # Another way to get new 'natural' bins
+        new_bins = [get_previous_date(date_bins[0], step)]+[get_next_date(date, step)-datetime.timedelta(seconds=1) for date in date_bins[1:-1]]+[get_next_date(date_bins[-1], step)]
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:446]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 19]
```diff
@@ -0,0 +1,535 @@
...
+        labels (List[str]) : The names to give to the bins
+        ratio_min (float) : The minimum ratio wanted between the less populated category over the most populated
+    Returns:
+        pd.Series : The binned column
+        bool : If the ratio condition is satisfied
+    '''
+    new_col = pd.cut(date_column, bins=new_bins, labels=labels, include_lowest=True)
+    count = new_col.value_counts()
+    ratio = min(count)/max(count)
+    return new_col, ratio>=ratio_min
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:473]: Pep8

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 20]
```diff
@@ -0,0 +1,535 @@
...
+def bin_datetime_col(data: pd.DataFrame, col: str, nb_bins: int) -> pd.Series:
+    '''Bins a date column in nb_bins
+
+    Args:
+        data (pd.DataFrame) : The dataset we consider
+        col (str) : The name of the date column to bin
+        nb_bins (int) : Number of bins we want
+    Returns:
+        pd.Series : The binned column
+
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:485]: Space

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 21]
```diff
@@ -0,0 +1,535 @@
...
+def normalize_data(data: pd.DataFrame, sensitive_cols: List[str], nb_bins: int = 5) -> pd.DataFrame:
+    '''Casts each sensitive col in a suitable dtype and bins them
+
+    Args:
+        data (pd.DataFrame) : The dataframe whose columns we want to cast
+        sensitive_cols (List[str]) : The list of the columns containing sensitive attributes (eg. sex, age, ethnicity,...)
+        nb_bins (int) : The number of bins to consider when binning a datetime or continuous column
+    Returns:
+        pd.DataFrame : The dataframe with the columns casted and binned
+
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:505]: Space

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 22]
```diff
@@ -0,0 +1,535 @@
...
+        nb_bins (int) : The number of bins to consider when binning a datetime or continuous column
+    Returns:
+        pd.DataFrame : The dataframe with the columns casted and binned
+    
+    '''
+    new_data = data.copy()
+    for col in sensitive_cols:
+        new_data[col] = fl_utils.infer_dtype(new_data[col])
+        distr_type = fl_utils.infer_distr_type(new_data[col])
+        if distr_type.value=='continuous':
```
Comment (LexABzH) [inline gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:511]: Pep8

File: gabarit/template_num/num_tests/functional_tests.py [hunk 23]
```diff
@@ -182,7 +182,109 @@ def test04_GenerateReport(self):
...
 
-    def test05_PreProcessData(self):
+    def test05_fairness_metrics(self):
+        '''Test of the file utils/0_fairness_metrics.py'''
+        print("Test of the file utils/0_fairness_metrics.py")
+
+        base_filenames_fairlens = ['data_biased_groups.csv', 'data_distribution_score.csv', 'data_distributions.png']
+        set_columns_data_distribution = {'Group', 'Distance', 'Proportion', 'Counts', 'P-Value'}
+        set_columns_data_biased = {'Group', 'Distance', 'Proportion', 'Counts', 'P-Value', 'number_of_attributes'}
+
```
Comment (LexABzH) [inline gabarit/template_num/num_tests/functional_tests.py:192]: Space ?

File: gabarit/template_num/num_tests/functional_tests.py [hunk 24]
```diff
@@ -182,7 +182,109 @@ def test04_GenerateReport(self):
...
+                remove_dir(output_path)
+
+            # Constitutes the list of files that should be present
+            if target == 'biased_target_binary_int':
+                list_metrics = list_metrics_binary
+            if target == 'biased_target':
+                list_metrics = list_metrics_continuous
+            if target == 'biased_target_str':
+                list_metrics = list_metrics_categorical
+            filenames = ['fairness_algo_barplot_'+metric+'.png' for metric in list_metrics]
```
Comment (LexABzH) [inline gabarit/template_num/num_tests/functional_tests.py:212]: Pep8

File: gabarit/template_num/num_tests/functional_tests.py [hunk 25]
```diff
@@ -182,7 +182,109 @@ def test04_GenerateReport(self):
...
+            if 'age' in sensitive_cols:
+                nb_groups_whole = nb_groups_whole * (nb_bins+1)
+                nb_groups_intersection = nb_groups_intersection * nb_bins
+            if 'citizenship' in sensitive_cols:
+                nb_groups_whole = nb_groups_whole * (2+1)
+                nb_groups_intersection = nb_groups_intersection * 2
+            if 'sex_label' in sensitive_cols:
+                nb_groups_whole = nb_groups_whole * (2+1)
+                nb_groups_intersection = nb_groups_intersection * 2
+            nb_groups_whole = nb_groups_whole-1
```
Comment (LexABzH) [inline gabarit/template_num/num_tests/functional_tests.py:233]: Pep8

**Review by LexABzH — COMMENTED** [review]
- Maybe avoid 'sex' label ? Use 'genre' ?
- In `0_fairness_metrics.py`, I would reorder the function definitions.

File: gabarit/template_num/num_project/package_name/monitoring/mlflow_logger.py [hunk 26]
```diff
@@ -219,10 +221,19 @@ def log_text(self, text: str, artifact_file: str) -> None:
 
         Args:
             text (str): A text
-            artifact_file (str): The run-relative artifact file path in posixpath format to which the dictionary is saved
+            artifact_file (str): The run-relative artifact file path in posixpath format to which the text is saved
         '''
         mlflow.log_text(text=text, artifact_file=artifact_file)
 
+    def log_figure(self, figure: matplotlib.figure.Figure, artifact_file: str) -> None:
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name/monitoring/mlflow_logger.py:228]: Done

File: gabarit/template_num/num_project/package_name/monitoring/mlflow_logger.py [hunk 27]
```diff
@@ -25,6 +25,8 @@
 import mlflow
 import pathlib
 import logging
+import matplotlib
+import matplotlib.pyplot as plt
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name/monitoring/mlflow_logger.py:29]: Yes. I don't know why, but in order to do the type hinting on the matplotlib figure, both are needed

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 28]
```diff
@@ -0,0 +1,535 @@
...
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:18]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 29]
```diff
@@ -0,0 +1,535 @@
...
+
+
+import os
+import logging
+import sklearn
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:23]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 30]
```diff
@@ -0,0 +1,535 @@
...
+import datetime
+import argparse
+import fairlearn
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:26]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 31]
```diff
@@ -0,0 +1,535 @@
...
+import matplotlib
+import pandas as pd
+import fairlens as fl
+import fairlearn.metrics
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:30]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 32]
```diff
@@ -0,0 +1,535 @@
...
+from functools import partial
+import matplotlib.pyplot as plt
+from typing import List, Union, Tuple
+from fairlens import utils as fl_utils
+
+from {{package_name}} import utils
+from {{package_name}}.monitoring.mlflow_logger import MLflowLogger
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:37]: To be discussed

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 33]
```diff
@@ -0,0 +1,535 @@
...
+    the number of attributes defining the group.
+
+    Args:
+        distribution_score (pd.DataFrame) : A dataframe obtained by the method distribution_score of a fl.FairnessScorer
+        min_proportion (float) : The minimal proportion of a subgroup to be considered as biased
+        min_distance (float) : The minimal distance (Kolmogorov-Smirnov) of a subgroup to be considered as biased
+        max_p_value (float) : The maximal p-value (Kolmogorov-Smirnov) of a subgroup to be considered as biased
+    Returns:
+        pd.DataFrame : the biased groups
+
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:106]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 34]
```diff
@@ -0,0 +1,535 @@
...
+    Returns:
+        pd.DataFrame : the biased groups
+    
+    '''
+    biased_groups = distribution_score.copy()
+    biased_groups = biased_groups[biased_groups['Proportion']>=min_proportion]
+    biased_groups = biased_groups[abs(biased_groups['Distance'])>=min_distance]
+    biased_groups = biased_groups[biased_groups['P-Value']<=max_p_value]
+    biased_groups = biased_groups.sort_values('Distance', ascending=False)
+    biased_groups['number_of_attributes'] = biased_groups['Group'].apply(lambda x:x.count(',')+1)
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:113]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 35]
```diff
@@ -0,0 +1,535 @@
...
+        get_fairlearn_metrics(data=data, col_target=col_target, col_pred=col_pred, sensitive_cols=sensitive_cols, 
+                            output_path=output_path, sep=sep, encoding=encoding, mlflow_logger=mlflow_logger)
+    if mlflow_logger is not None:
+        mlflow_logger.end_run()
+
+
+def find_bias(distribution_score: pd.DataFrame, 
+              min_proportion: float, 
+              min_distance: float, 
+              max_p_value: float) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:94]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 36]
```diff
@@ -0,0 +1,535 @@
...
+        elif distr_type.value=='datetime':
+            new_data[col] = bin_datetime_col(new_data, col, nb_bins)
+    return new_data
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser('fairness_metrics', description=(
+            "Calculates various metrics for fairness."
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:522]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 37]
```diff
@@ -0,0 +1,535 @@
...
+        col_target (str) : The name of the target column in data
+        sensitive_cols (List[str]) : The list of the columns containing sensitive attributes (eg. sex, age, ethnicity,...)
+        output_path (str) : The path to the folder where the files will be saved
+    Kwargs:
+        sep (str): Separator to use with the .csv files
+        encoding (str): Encoding to use with the .csv files
+        mlflow_logger (MLflowLogger) : The logger to log the metrics in MLflow
+    '''
+    # Instanciates the fl.FairnessScorer
+    fl_scorer = fl.FairnessScorer(data[sensitive_cols+[col_target]].copy(), col_target, sensitive_attrs = sensitive_cols)
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:136]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py [hunk 38]
```diff
@@ -0,0 +1,535 @@
...
+    # Plots and saves the distributions
+    logger.info(f"Calculates distributions graphs")
+    fl_scorer.plot_distributions(normalize=True)
+    fig_distributions = plt.gcf()
+    plt.savefig(os.path.join(output_path, 'data_distributions.png'), bbox_inches="tight")
+    # Calculates and saves the Kolmogorov-Smirnov distances
+    logger.info(f"Calculates Kolmogorov-Smirnov distances for each subgroup")
+    for col in fl_scorer.sensitive_attrs:
+        fl_scorer.df[col] = fl_scorer.df[col].apply(lambda x:f'({col}) '+str(x))
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:145]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py (hunk 1 above)
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:170]: Done

File: gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py (hunk 2 above)
Comment (gsolard) [reply gabarit/template_num/num_project/package_name-scripts/utils/0_fairness_metrics.py:223]: Done

File: gabarit/template_nlp/nlp_tests/functional_tests.py [hunk 39]
```diff
@@ -154,26 +154,30 @@ def test04_SplitTrainValidTest(self):
         self.assertFalse(any([_ in df_valid.x_col.values for _ in df_test.x_col.values]))
 
     def test05_GenerateReport(self):
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_tests/functional_tests.py:156]: Rename function

File: gabarit/template_nlp/nlp_tests/functional_tests.py [hunk 40]
```diff
@@ -154,26 +154,30 @@ def test04_SplitTrainValidTest(self):
...
         if os.path.exists(config_path):
             os.remove(config_path)
         with open(config_path, 'w') as f:
             json.dump({"open_browser": False}, f)
+        report_path = os.path.join(full_path_lib, "test_template_nlp-data", "reports", "sweetviz")
+        remove_dir(report_path)
 
         # "Basic" case
-        basic_run = f"{activate_venv}python {full_path_lib}/test_template_nlp-scripts/utils/0_generate_report.py -s mono_class_mono_label.csv --source_names source --config {config_path}"
+        basic_run = f"{activate_venv}python {full_path_lib}/test_template_nlp-scripts/utils/0_sweetviz_report.py -s mono_class_mono_label.csv --source_names source --config {config_path}"
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_tests/functional_tests.py:170]: Add a test with mlflow ?  --mlflow_experiment gabarit_ci/mlflow_test ?

File: gabarit/template_num/num_tests/functional_tests.py [hunk 41]
```diff
@@ -145,26 +145,30 @@ def test03_SplitTrainValidTest(self):
         self.assertEqual(df_test.shape[0], 42)
 
     def test04_GenerateReport(self):
```
Comment (LexABzH) [inline gabarit/template_num/num_tests/functional_tests.py:147]: Rename function

File: gabarit/template_num/num_tests/functional_tests.py [hunk 42]
```diff
@@ -145,26 +145,30 @@ def test03_SplitTrainValidTest(self):
...
         if os.path.exists(config_path):
             os.remove(config_path)
         with open(config_path, 'w') as f:
             json.dump({"open_browser": False}, f)
+        report_path = os.path.join(full_path_lib, "test_template_num-data", "reports", "sweetviz")
+        remove_dir(report_path)
 
         # "Basic" case
-        basic_run = f"{activate_venv}python {full_path_lib}/test_template_num-scripts/utils/0_generate_report.py -s mono_class_mono_label.csv --source_names source --config {config_path}"
+        basic_run = f"{activate_venv}python {full_path_lib}/test_template_num-scripts/utils/0_sweetviz_report.py -s mono_class_mono_label.csv --source_names source --config {config_path}"
```
Comment (LexABzH) [inline gabarit/template_num/num_tests/functional_tests.py:161]: Add a test with mlflow ? --mlflow_experiment gabarit_ci/mlflow_test ?

File: gabarit/template_num/num_tests/functional_tests.py [hunk 43]
```diff
@@ -175,14 +179,122 @@ def test04_GenerateReport(self):
...
+            if 'citizenship' in sensitive_cols:
+                nb_groups_whole = nb_groups_whole * (2 + 1)
+                nb_groups_intersection = nb_groups_intersection * 2
+            if 'genre' in sensitive_cols:
+                nb_groups_whole = nb_groups_whole * (2 + 1)
+                nb_groups_intersection = nb_groups_intersection * 2
+            nb_groups_whole = nb_groups_whole - 1
+
+            # Run the script
+            basic_run = f"{activate_venv}python {full_path_lib}/test_template_num-scripts/utils/0_fairness_report.py -f {filename_test_fairness} -t {target} -n {nb_bins} -s"
```
Comment (LexABzH) [inline gabarit/template_num/num_tests/functional_tests.py:240]: Add a test with mlflow ? --mlflow_experiment gabarit_ci/mlflow_test ?

**Review by LexABzH — COMMENTED** [review]
OK,

Shouldn't we check that report files are created in the mlflow local folder in functional tests ?

**Comment by gsolard** [comment]
I opened an issue #86 on this matter

**Pull request merged**